cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/DistributedProgramRunner.java

```diff
@@ -591,17 +593,25 @@ private ProgramOptions updateProgramOptions(ProgramOptions options,
         BundleJarUtil.addToArchive(localDir, jarOut);
       }
 
+      String pluginDirHash = "";
+      if (systemArgs.hasOption(ProgramOptionConstants.PLUGIN_DIR_HASH)) {
+        // if hash value for plugins has been provided, we append it to filename.
+        pluginDirHash = systemArgs.getOption(ProgramOptionConstants.PLUGIN_DIR_HASH);
+        pluginDirFileName = String.format("%s_cache_%s", PLUGIN_DIR, pluginDirHash);
+        pluginArchiveFileName = PLUGIN_ARCHIVE.replace(".jar", String.format("_cache_%s%s", pluginDirHash, ".jar"));
```

Comment:
Can you add a code comment about how the caching directory is being organized?

Reply:
The goal at this point is to add `_cache_` to an artifact file name. This states that the artifact is cacheable. Particular provisioner is then responsible to leverage this information and organize caches. 
Already added [comments](https://github.com/cdapio/cdap/pull/14521/files#diff-b7675d51d200ad4368d96b266bc04c96d4ecd20e5d1e1a6510f43fee970ffacbR190) on how cacheable artifacts are organized in Dataproc:

Reply:
I'll add comment here to make it clear.

Reply:
This seems a bit hacky to base on pattern matching on the name. It's better to encapsulate the information into the `LocalFile` instance. We can introduce a tagging interface in the runtime-spi (e.g. `Cacheable`) such that the provisioner implementation can use `instanceof` to check.

Reply:
what about using an interface that has a getHash() method instead of modifying the filename? Then leave it up to the caller to decide how to use it? For example, dataproc provisioner could decide to attach a hash label to theh object instead of relying on filename there

Reply:
Discussed offline with Terence on reasons why we cannot use LocalFile implementation in DistributedProgramRunner (main reason is that we use LocalizeResource which only has URI and LocalFiles are initialized in twill part later on).
We also need to modify file name since file names are then added to `program.options.json`.

Based on our discussion, we agreed to (1) avoid exposing hash to other extensions (e.g., Dataproc) and (2) use a new CacheableLocalFile in runtime-spi. So now it's up to the provisioner to decide to cache or not.